File: sonar-go-checks/src/test/resources/checks/SwitchWithoutDefaultCheck/SwitchWithoutDefaultCheck.go
package SwitchWithoutDefaultCheck

func main() {
    switch x { // Noncompliant {{Add a default clause to this "switch" statement.}}
//  ^^^^^^
    }

    switch x { // Noncompliant {{Add a default clause to this "switch" statement.}}
//  ^^^^^^
    case 1:
      fmt.Println("1")
    }

    switch x {
    case 1:
      fmt.Println("1")
    default:
      fmt.Println("default")
    }

    switch x {
    default:
      fmt.Println("default")
    case 1:
      fmt.Println("1")
    }

    switch x {
    case 1:
      fmt.Println("1")
    default:
      switch y { // Noncompliant {{Add a default clause to this "switch" statement.}}
   // ^^^^^^
      case 2:
        fmt.Println("2")
      }
    }
}

File: sonar-go-checks/src/main/java/org/sonar/go/checks/SwitchWithoutDefaultCheck.java
/*
 * SonarSource Go
 * Copyright (C) 2018-2025 SonarSource Sàrl
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource Sàrl.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
package org.sonar.go.checks;

import org.sonar.check.Rule;
import org.sonar.plugins.go.api.MatchTree;
import org.sonar.plugins.go.api.checks.GoCheck;
import org.sonar.plugins.go.api.checks.InitContext;

@Rule(key = "S131")
public class SwitchWithoutDefaultCheck implements GoCheck {

  @Override
  public void initialize(InitContext init) {
    init.register(MatchTree.class, (ctx, tree) -> {
      if (tree.cases().stream().noneMatch(matchCase -> matchCase.expression() == null)) {
        var keyword = tree.keyword();
        var message = String.format("Add a default clause to this \"%s\" statement.", keyword.text());
        ctx.reportIssue(keyword, message);
      }
    });
  }

}
